File: book_catalogue/services/open_library/schemas/author.py
from __future__ import annotations
__all__ = ["Author"]

from pydantic import Field

from book_catalogue.services.open_library.schemas import (
    BaseModel,
    DatetimeResource,
    Resource,
    TextResource,
)


class Identifiers(BaseModel):
    amazon: str | None = None
    goodreads: str | None = None
    isni: str | None = None
    librarything: str | None = None
    viaf: str | None = None
    wikidata: str | None = None


class Link(BaseModel):
    title: str
    type: Resource
    url: str


class Author(BaseModel):
    alternate_names: list[str] = Field(default_factory=list)
    bio: str | TextResource | None = None
    birth_date: str | None = None
    created: DatetimeResource | None = None
    death_date: str | None = None
    entity_type: str | None = None
    id: int | None = None
    key: str
    last_modified: DatetimeResource
    latest_revision: int | None = None
    links: list[Link] = Field(default_factory=list)
    name: str
    personal_name: str | None = None
    photos: list[int] = Field(default_factory=list)
    remote_ids: Identifiers = Identifiers()
    revision: int
    source_records: list[str] = Field(default_factory=list)
    title: str | None = None
    type: Resource
    wikipedia: str | None = None

    @property
    def author_id(self) -> str:
        return self.key.split("/")[-1]
    
    def get_bio(self) -> str | None:
        if self.bio:
            if isinstance(self.bio, TextResource):
                return self.bio.value
            return self.bio
        return None

File: openlibrary/src/main/kotlin/github/buriedincode/openlibrary/schemas/Author.kt
package github.buriedincode.openlibrary.schemas

import github.buriedincode.openlibrary.serializers.DescriptionSerializer
import github.buriedincode.openlibrary.serializers.LocalDateSerializer
import github.buriedincode.openlibrary.serializers.LocalDateTimeSerializer
import kotlinx.datetime.LocalDate
import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.Serializable

@Serializable
data class Author(
    val alternateNames: List<String> = emptyList(),
    @Serializable(with = DescriptionSerializer::class)
    val bio: String? = null,
    val birthDate: String? = null,
    val comment: String? = null,
    @Serializable(with = LocalDateTimeSerializer::class)
    val created: LocalDateTime? = null,
    val date: String? = null,
    @Serializable(with = LocalDateSerializer::class)
    val deathDate: LocalDate? = null,
    val entityType: String? = null,
    val fullerName: String? = null,
    val id: Long? = null,
    val key: String,
    @Serializable(with = LocalDateTimeSerializer::class)
    val lastModified: LocalDateTime?,
    val latestRevision: Int? = null,
    val links: List<Link> = emptyList(),
    val location: String? = null,
    val name: String,
    val personalName: String? = null,
    val photos: List<Int> = emptyList(),
    val remoteIds: RemoteIds? = null,
    val revision: Int,
    val sourceRecords: List<String> = emptyList(),
    val title: String? = null,
    val type: Resource,
    val wikipedia: String? = null,
) {
    @Serializable
    data class RemoteIds(
        val amazon: String? = null,
        val goodreads: String? = null,
        val isni: String? = null,
        val librarything: String? = null,
        val librivox: String? = null,
        val projectGutenberg: String? = null,
        val storygraph: String? = null,
        val viaf: String? = null,
        val wikidata: String? = null,
    )
}
